### PR TITLE
Adding CC310 entropy driver utilizing nrf_cc310_platform library

### PR DESCRIPTION
**Pull request by frkv**
-Adding hw_cc310 driver with RTOS mutex/abort support
-Changing entropy_cc310 driver to use nrf_cc310_platform library
-Changing entropy_cc310 driver to support non-secure application

Added description for:
- cc310 entropy driver
- cc310 hw initialization

Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

### CODE REVIEW

**Comment by tejlmand** [comment]
@rlubos Please check this commit: https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1268/commits/e4ecdef502d7b3ccdd1b11c4719888bf88de6664

**Comment by pdunaj** [comment]
Hi @frkv , can you tell why it is disabled for desktop? Is there any reason we would like to have it enabled?

**Comment by tejlmand** [comment]
> Hi @frkv , can you tell why it is disabled for desktop? Is there any reason we would like to have it enabled?

When enabling cc310 on nrf desktop, we face the following issue:
```
Found 12 asserts!
desktop52_base_test Line:223  Assert raised, [00000003] <err> os: ***** MPU FAULT *****
desktop52_base_test Line:223  Assert raised, [00000003] <err> os:   Stacking error (context area might be not valid)
desktop52_base_test Line:223  Assert raised, [00000003] <err> os:   Data Access Violation
desktop52_base_test Line:223  Assert raised, [00000003] <err> os:   MMFAR Address: 0x2000829c
desktop52_base_test Line:223  Assert raised, [00000004] <err> os: r0/a1:  0x00000000  r1/a2:  0x0002be03  r2/a3:  0x00000000
desktop52_base_test Line:223  Assert raised, [00000004] <err> os: r3/a4:  0x0002be03 r12/ip:  0x0004134a r14/lr:  0x01000000
desktop52_base_test Line:223  Assert raised, [00000004] <err> os:  xpsr:  0x20001cbc
desktop52_base_test Line:223  Assert raised, [00000004] <err> os: Faulting instruction address (r15/pc): 0x2000cd38
desktop52_base_test Line:223  Assert raised, [00000004] <err> os: >>> ZEPHYR FATAL ERROR 2: Stack overflow
desktop52_base_test Line:223  Assert raised, [00000004] <err> os: >>> ZEPHYR FATAL ERROR 2: Stack overflow
desktop52_base_test Line:223  Assert raised, [00000005] <err> os: Current thread: 0x20005734 (unknown)
desktop52_base_test Line:223  Assert raised, [00328145] <err> os: Halting system
```
which indicates that MPU must be configured differently.

But as first step, the cc310 is disabled in nrf desktop, as this should be identical to how nrf desktop is working without this PR.
Then it allows time to look into the MPU configuration aftwerwards.

**Comment by pdunaj** [comment]
Hi @tejlmand , nrf_desktop uses the default MPU configuration from Zephyr. What we use however in the debug build is stack guard ({{CONFIG_MPU_STACK_GUARD}}). Note that this may be something that others will use as well so it would be good if changes introduced are compatible with this feature. Or if the feature is not compatible it would be good to set your config to depend on {{!MPU_STACK_GUARD}}.

**Comment by tejlmand** [comment]
> Hi @tejlmand , nrf_desktop uses the default MPU configuration from Zephyr. What we use however in the debug build is stack guard (`CONFIG_MPU_STACK_GUARD`). Note that this may be something that others will use as well so it would be good if changes introduced are compatible with this feature. Or if the feature is not compatible it would be good to set your config to depend on `!MPU_STACK_GUARD`.

Thanks for info.
Has changed to `MPU_STACK_GUARD` dependency.
Of course this must work with the MPU in future, but this allows for introducing the entropy driver through SPM as a first step.

**Comment by pdunaj** [comment]
> > Hi @tejlmand , nrf_desktop uses the default MPU configuration from Zephyr. What we use however in the debug build is stack guard (`CONFIG_MPU_STACK_GUARD`). Note that this may be something that others will use as well so it would be good if changes introduced are compatible with this feature. Or if the feature is not compatible it would be good to set your config to depend on `!MPU_STACK_GUARD`.
> 
> Thanks for info.
> Has changed to `MPU_STACK_GUARD` dependency.
> Of course this must work with the MPU in future, but this allows for introducing the entropy driver through SPM as a first step.

Thanks. Also is it possible that the feature would not be enabled by default? This is to make sure that currently working configurations will be fine. Apps/samples that need it can enabled it in the config. Multiple things are not enabled by default, Bluetooth included.

**Comment by tejlmand** [comment]
> > > Hi @tejlmand , nrf_desktop uses the default MPU configuration from Zephyr. What we use however in the debug build is stack guard (`CONFIG_MPU_STACK_GUARD`). Note that this may be something that others will use as well so it would be good if changes introduced are compatible with this feature. Or if the feature is not compatible it would be good to set your config to depend on `!MPU_STACK_GUARD`.
> > 
> > 
> > Thanks for info.
> > Has changed to `MPU_STACK_GUARD` dependency.
> > Of course this must work with the MPU in future, but this allows for introducing the entropy driver through SPM as a first step.
> 
> Thanks. Also is it possible that the feature would not be enabled by default? This is to make sure that currently working configurations will be fine. Apps/samples that need it can enabled it in the config. Multiple things are not enabled by default, Bluetooth included.

If a sample requires entropy, then I would argue that we should aim for most secure way of obtaning that, and that would be to use the cc310 when it is available.
Just as you use the nrf5 rng today, because there were no driver for the cc310 entropy.
Note, only samples that sets `ENTROPY_GENERATOR` will use the cc310 entropy if available, so cc310 entropy is not enabled everyewhere.

File: samples/nrf9160/lwm2m_carrier/prj.conf [hunk 1]
```diff
@@ -9,9 +9,6 @@ CONFIG_LWM2M_CARRIER=y
 # The library requires newlibc
 CONFIG_NEWLIB_LIBC=y
 
-# The library needs a source of entropy
-CONFIG_TEST_RANDOM_GENERATOR=y
```
Comment (rlubos) [inline samples/nrf9160/lwm2m_carrier/prj.conf]: There are two `overlay-carrier.conf` files that should be updated as well in mqtt and coap samples.

Comment (tejlmand) [reply samples/nrf9160/lwm2m_carrier/prj.conf]: Fixed

**Comment by rlubos** [comment]
@b-gent Could have a look at docs in this PR?

**Comment by b-gent** [comment]
If the CI is green, let us merge now and I will start reviewing the docs on my own branch today.

**Comment by b-gent** [comment]
``nrf/rst/doc/nrf/drivers.rst:10: WARNING: toctree glob pattern '../../doc/drivers/*' didn't match any documents``

@tejlmand

**Comment by tejlmand** [comment]
> ivers.rst:10: WARNING:

strange, that worked locally
looking at it.

**Comment by tejlmand** [comment]
dox commit dropped, as not needed for this PR.
Will make separate PR with dox content when this is merged

**Pull request merged**